### ISSUE TITLE
Traffic light feature requset

### ISSUE THREAD

**Issue by HIGHGO77** [CitiesSkylinesMods/TMPE#1190]
### Describe your idea

The signal lamp controlling the entrance of a vehicle in one direction should be at the exit of the intersection in front of the vehicle, but it is actually located on the left side of the entrance of the intersection, which is not in line with the actual traffic control situation. In addition, because only one set of light color channel is open for signal control, the left, straight and right direction light panel can only be displayed as red or green at the same time, and the correct left-turn matching light color cannot be realized.

控制一个方向进口车辆的信号灯应该在车辆前方的路口出口处，但实际上却位于路口进口道的左边，这不符合实际的交通控制状况。另外，由于信号灯控只开放了一组灯色通道，导致左、直、右三方向灯盘只能同时显示为红色或绿色，无法实现正确的左转对放灯色。

### Screenshots?

![Uploading 屏幕截图 2021-11-09 170456.png…]()

**Comment by krzychu124** [comment]
> The signal lamp controlling the entrance of a vehicle in one direction should be at the exit of the intersection in front of the vehicle, but it is actually located on the left side of the entrance of the intersection, which is not in line with the actual traffic control situation

Out of Traffic Manager mod scope. Use different set of assets from the workshop to achieve that setting: [example asset pack](https://steamcommunity.com/sharedfiles/filedetails/?id=1535107168) As far as I know all American traffic light packs look/work like you've described.  

> In addition, because only one set of light color channel is open for signal control, the left, straight and right direction light panel can only be displayed as red or green at the same time, and the correct left-turn matching light color cannot be realized.

I assume you suggest that traffic light prop should light up part/separate light of the traffic light to match with what TrafficManager is showing. That requires designing completely new shader for traffic light props, and probably pretty advanced mod/addon/extension to manage separate lights of traffic lights props. 
I've been experimenting with that, but I don't have time to push it forward and also task by itself is really complex. Not to mention that new shader and traffic light manager would require completely new set of traffic light props that can utilize new shader and "communicate" with mentioned addon/extension. Also Traffic Manager cannot control traffic light state per lane yet so a lot of work must be done before.

As of today, out of scope, could work as an addon to Traffic Manager.

**Comment by HIGHGO77** [comment]
Thank you for your response.

**Comment by HIGHGO77** [comment]
Thank you for your response.



















At 2021-11-10 03:59:45, "Krzysztof" ***@***.***> wrote:

The signal lamp controlling the entrance of a vehicle in one direction should be at the exit of the intersection in front of the vehicle, but it is actually located on the left side of the entrance of the intersection, which is not in line with the actual traffic control situation

Out of Traffic Manager mod scope. Use different set of assets from the workshop to achieve that setting: example asset pack As far as I know all American traffic light packs look/work like you've described.

In addition, because only one set of light color channel is open for signal control, the left, straight and right direction light panel can only be displayed as red or green at the same time, and the correct left-turn matching light color cannot be realized.

I assume you suggest that traffic light prop should light up part/separate light of the traffic light to match with what TrafficManager is showing. That requires designing completely new shader for traffic light props, and probably pretty advanced mod/addon/extension to manage separate lights of traffic lights props.
I've been experimenting with that, but I don't have time to push it forward and also task by itself is really complex. Not to mention that new shader and traffic light manager would require completely new set of traffic light props that can utilize new shader and "communicate" with mentioned addon/extension. Also Traffic Manager cannot control traffic light state per lane yet so a lot of work must be done before.

As of today, out of scope, could work as an addon to Traffic Manager.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by aubergine10** [comment]
Actually, I think some of this is possible, although out-of-scope of TMPE.

For placing traffic light on other side of junction, that can be achieved with recently released [Traffic Lights Replacer](https://steamcommunity.com/linkfilter/?url=https://tlr.cgameworld.com/documentation/pack-creation/) mod - see mod description for details.

And for the left turn indicator, is it for **Turn on Red** policy? If so, remember that traffic lights additionally have pedestrian lights which can be used - eg. when traffic light is red, pedestrian "cross road" light turns green. I have replaced traffic lights on one of my roads with a traffic light prop which works that way - it lights "turn left" indicator (actually pedestrian "cross road" indicator repurposed as "turn left" indicator) when the main traffic lights turn red ([as described here](https://github.com/CitiesSkylinesMods/TMPE/wiki/Turn-on-Red)):

![Allow Turn Right on Red](https://imgur.com/GYo4GqF.png) ![Deny Turn Right on Red](https://imgur.com/BxJDYLT.png)

There are limits, for example currently traffic light replacement is per road prefab so I can't just do this at specific junctions - it's every traffic light on that road prefab, or none. Although I can of course subscribe several similar road prefabs to work around that (bit clunky, but viable).

And the other limit, of course, is that traffic light props have no concept of lanes; so we can't individually control a light per lane which is really annoying. The whole road is either "green" or "red" as far as the prop is concerned.
